libvast/src/system/eraser.cpp

```diff
@@ -0,0 +1,88 @@
+/******************************************************************************
+ *                    _   _____   __________                                  *
+ *                   | | / / _ | / __/_  __/     Visibility                   *
+ *                   | |/ / __ |_\ \  / /          Across                     *
+ *                   |___/_/ |_/___/ /_/       Space and Time                 *
+ *                                                                            *
+ * This file is part of VAST. It is subject to the license terms in the       *
+ * LICENSE file found in the top-level directory of this distribution and at  *
+ * http://vast.io/license. No part of VAST, including this file, may be       *
+ * copied, modified, propagated, or distributed except according to the terms *
+ * contained in the LICENSE file.                                             *
+ ******************************************************************************/
+
+#include "vast/system/eraser.hpp"
+
+#include <caf/event_based_actor.hpp>
+#include <caf/stateful_actor.hpp>
+#include <caf/timespan.hpp>
+
+#include "vast/concept/parseable/to.hpp"
+#include "vast/concept/parseable/vast/expression.hpp"
+#include "vast/expression.hpp"
+#include "vast/logger.hpp"
+#include "vast/system/atoms.hpp"
+
+namespace vast::system {
+
+eraser_state::eraser_state(caf::event_based_actor* self) : super(self) {
+  // nop
+}
+
+void eraser_state::init(caf::timespan interval, std::string query,
+                        caf::actor index_hdl, caf::actor archive_hdl) {
+  // Set member variables.
+  interval_ = interval;
+  query_ = std::move(query);
+  index_ = std::move(index_hdl);
+  archive_ = std::move(archive_hdl);
+  // Override the behavior for the idle state.
+  behaviors_[idle].assign([=](run_atom) {
+    auto expr = to<expression>(query_);
+    if (!expr) {
+      VAST_ERROR(self_, "failed to parse query", query_);
+      return;
+    }
```

Comment:
This should probably be an assertion here; and user input already verified at an earlier point.

Reply:
I generally prefer a defensive programming style that makes no assumption on the behavior of upstream components. This is potentially user-provided input, so we should treat it with care. It's not an assertion to me, because malformed input here only affects this local component and not the integrity of the entire system.